#! /bin/sh -e

# DP: Revert rgbimgmodule change in revision 60793.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "usage: `basename $0`: -patch|-unpatch [-d <srcdir>]"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
	echo >&2 "usage: `basename $0`: -patch|-unpatch [-d <srcdir>]"
        exit 1
esac
exit 0

--- Modules/rgbimgmodule.c.orig	2008-08-19 20:57:46.000000000 +0000
+++ Modules/rgbimgmodule.c	2010-11-28 16:17:39.860647406 +0000
@@ -269,7 +269,7 @@
 	Py_Int32 *starttab = NULL, *lengthtab = NULL;
 	FILE *inf = NULL;
 	IMAGE image;
-	int y, z, tablen, new_size;
+	int y, z, tablen;
 	int xsize, ysize, zsize;
 	int bpp, rle, cur, badorder;
 	int rlebuflen;
@@ -306,15 +306,9 @@
         }
 	if (rle) {
 		tablen = ysize * zsize * sizeof(Py_Int32);
-		rlebuflen = (int) (1.05 * xsize +10);
-		if ((tablen / sizeof(Py_Int32)) != (ysize * zsize) ||
-		    rlebuflen < 0) {
-			PyErr_NoMemory();
-			goto finally;
-		}
-
 		starttab = (Py_Int32 *)malloc(tablen);
 		lengthtab = (Py_Int32 *)malloc(tablen);
+		rlebuflen = (int) (1.05 * xsize +10);
 		rledat = (unsigned char *)malloc(rlebuflen);
 		if (!starttab || !lengthtab || !rledat) {
 			PyErr_NoMemory();
@@ -342,14 +336,8 @@
 
 		fseek(inf, 512 + 2 * tablen, SEEK_SET);
 		cur = 512 + 2 * tablen;
-		new_size = xsize * ysize + TAGLEN;
-		if (new_size < 0 || (new_size * sizeof(Py_Int32)) < 0) {
-			PyErr_NoMemory();
-			goto finally;
-		}
-
 		rv = PyString_FromStringAndSize((char *)NULL,
-				      new_size * sizeof(Py_Int32));
+				      (xsize * ysize + TAGLEN) * sizeof(Py_Int32));
 		if (rv == NULL)
 			goto finally;
 
@@ -417,14 +405,8 @@
 			copybw((Py_Int32 *) base, xsize * ysize);
 	}
 	else {
-		new_size = xsize * ysize + TAGLEN;
-		if (new_size < 0 || (new_size * sizeof(Py_Int32)) < 0) {
-			PyErr_NoMemory();
-			goto finally;
-		}
-
 		rv = PyString_FromStringAndSize((char *) 0,
-						new_size*sizeof(Py_Int32));
+					   (xsize*ysize+TAGLEN)*sizeof(Py_Int32));
 		if (rv == NULL)
 			goto finally;
 
@@ -614,16 +596,10 @@
 		return NULL;
 	}
 	tablen = ysize * zsize * sizeof(Py_Int32);
-	rlebuflen = (int) (1.05 * xsize + 10);
-
-	if ((tablen / sizeof(Py_Int32)) != (ysize * zsize) ||
-	    rlebuflen < 0 || (xsize * sizeof(Py_Int32)) < 0) {
-		PyErr_NoMemory();
-		goto finally;
-	}
 
 	starttab = (Py_Int32 *)malloc(tablen);
 	lengthtab = (Py_Int32 *)malloc(tablen);
+	rlebuflen = (int) (1.05 * xsize + 10);
 	rlebuf = (unsigned char *)malloc(rlebuflen);
 	lumbuf = (unsigned char *)malloc(xsize * sizeof(Py_Int32));
 	if (!starttab || !lengthtab || !rlebuf || !lumbuf) {
